design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,150 @@
+connections. But the circuit breaking functionality itself won't be protocol
+agnostic. Its implementation would be protocol specific and it's upto community
+which protocol they want to add support for and how circuit breaking will make
+sense in that protocol.
+
+## Design proposal
+
+[design-proposal]: #design-proposal
+
+Right now if I just consider ServiceProfile spec, IMO it seems to represent more
+of profile-for-http-service rather than profile-for-generic-service. May be it
+makes sense given that it supports only HTTP and grpc over HTTP as of now. Given
+current architecture, I feel it's best to provide circuit breaking functionality
+through service profiles. That means it would be limited to above two kind of
+services as of now.
+
+For TCP connections, according to original github
+[issue](https://github.com/linkerd/linkerd2/issues/2846#issue-447856936), it
+seems like linkerd2 does support automatic circuit breaking for TCP errors but
+I'm not sure of its behaviour -- whether TCP error means connection refused or
+resets or timeouts or something else. Anyway, it seems like it's not
+configurable and I'm not sure if making it configurable has any important
+usecases.
+
+In future, for any other protocol support for circuit breaking, one will first
+need to incorporate the protocol into ServiceProfile itself and then implement
+the circuit breaking functionality.
+
+Regarding ejection of the service instance, once we decide any instance needs to
+be ejected in accordance with mentioned configuration, it would be ejected from
+all the proxies i.e. we **won't** provide user ability to configure it just for
+proxy which raised those failures due to which ejection got triggered.
+
+Let me explain above piece with an example. Let's say there are two services A
+and B which calls C service. In case B->C gets success rate of 100% but A->C
+gets success rate of 50%, it won't be possible for user to configure
+service-mesh such that it would trigger circuit breaking only for A->C traffic.
+Instead either circuit breaking will happen for both A and B (through
+aggregation of their stats) or none of them.
+
+This certainly has its own disadvantages but I would be somewhat inclined
+towards having it simpler unless usecases for making it per-proxy over-weighs
+other one. So, for now, I propose that every ejection policy would be based on
+taking aggregated metrics over all the proxy instances to mark any instance for
+ejection.
+
+For HTTP/gRPC, I feel like taking advantage of linkerd provided golden metrics
+for ejection policies. Ejection can be based on success-rate,
+percentile-latencies for a given time window. Circuit breaker settings would be
+applied per service + protocol instead of per-route. In case circuit breaker is
+triggered for entire service, then we will return 503 for HTTP with some custom
+header and gRPC message for gRPC.
+
+If we go with the above design, may be we can keep most of the ejection logic in
```

Comment:
It is likely that each proxy will have a different view of the world (imagine a network split). Wouldn't it be better to have the logic inside each proxy, as they'll be the only ones that have a full picture of what's going on?